### PR TITLE
fix(processors.starlark): Use tracking ID to identify tracking metrics

### LINKED ISSUE #14484: "panic: negative refcount" when mutating tags in starlark for metrics from a tail input
### Relevant telegraf.conf

```toml
[[outputs.discard]]

[[inputs.tail]]
  files = ["/var/log/something.log"]
  data_format = "json"
  json_time_key = "Time"
  json_time_format = "unix"
  watch_method = "poll"

[[processors.starlark]]
  script = "/etc/telegraf/postprocess.star"
```

### Logs from Telegraf

```text
panic: negative refcount

goroutine 53 [running]:
github.com/influxdata/telegraf/metric.(*trackingMetric).decr(0xc002608f00?)
/go/src/github.com/influxdata/telegraf/metric/tracking.go:145 +0x85
github.com/influxdata/telegraf/metric.(*trackingMetric).Drop(0xc0021b9180?)
/go/src/github.com/influxdata/telegraf/metric/tracking.go:139 +0x13
github.com/influxdata/telegraf/agent.(*Agent).runAggregators.func1()
/go/src/github.com/influxdata/telegraf/agent/agent.go:720 +0xee
created by github.com/influxdata/telegraf/agent.(*Agent).runAggregators in goroutine 114
/go/src/github.com/influxdata/telegraf/agent/agent.go:707 +0x329
```

### System info

Telegraf >= 1.29.0

### Docker

Here is a minimal way to reproduce the error:

- `docker-compose.yaml`

    ```yaml
    version: "3"
    
    services:
      telegraf:
        image: telegraf:1.29
        volumes:
          - ./telegraf.conf:/etc/telegraf/telegraf.conf:ro
          - ./postprocess.star:/etc/telegraf/postprocess.star:ro
          - ./something.log:/var/log/something.log:ro
        command: --debug
    ```

- `./telegraf.conf` is the configuration file posted above
- `./postprocess.star`

    ```python
    def apply(metric):
        # Cloning the metric before mutating the tags *appears* to solve the issue
        # Would this have any unforeseen consequences? Is this a valid workaround?
        # metric = deepcopy(metric)
    
        print(metric)
        metric.tags.pop("path", None)

        # Other operations that cause the panic include setting tag values:
        # metrics.tags["path"] = "some-override"
    
        return [metric]
    ```

- `./something.log` should start as an empty file

### Steps to reproduce

1. Run the the above setup with `docker compose up`
2. Write to `something.log` using something like

    ```sh
    echo '{ "Value": 1, "Time": "'(date +%s)'" }' >> something.log
    ```

3. You'll see the output of the `print` line in the `starlark` showing the metric being processed
4. Wait until the metrics are flushed to the output — at that point you'll get a `panic: negative refcount` message and the docker container will exit

### Expected behavior

- No golang panics
- Metric tags should be able to be mutated in the `starlark` processor

### Actual behavior

- golang panics when metric tags are mutated in `starlark`

### Additional info

- This bug is only present from 1.29 onwards (you can easily see this by changing the `image` tag in the `docker-compose.yaml` file shown above)
- Seems to be closely related to https://github.com/influxdata/telegraf/issues/13148
- There may be a bunch of other metric mutation operations that also cause this panic but we have only been mutating tags
- This only seems to happen for `tail` inputs — `file` inputs don't suffer from the same issue

### PR DESCRIPTION
**Pull request by srebhan**
## Summary

To handle tracking-metrics correctly in the starlark processor, we need to identify tracking metrics and if they are "swallowed" by the processor.
Currently this process is based on comparing the returned metric hash-ID with the one if the input metric. However, this is fragile because the the hash-ID changes as soon as the metric is modified (e.g. renamed, tags are added/removed or tag-values are modified). As this is a very common use-case of the starlark processor, the present PR used the tracking-ID to identify returned tracking metrics. The PR furthermore adds  some more unit-tests for stateful processing scripts.

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #14484

### CODE REVIEW

File: plugins/processors/starlark/starlark_test.go [hunk 1]
```diff
@@ -3468,6 +3477,92 @@ def apply(metric):
...
+			// Do the requested number of loops
+			expected := make([]telegraf.TrackingID, 0, tt.loops)
+			for i := 0; i < tt.loops; i++ {
+				// Process expected metrics and compare with resulting metrics
+				input, tid := metric.WithTracking(testutil.TestMetric(i), notify)
+				expected = append(expected, tid)
+				require.NoError(t, plugin.Add(input, acc))
+			}
+			plugin.Stop()
+			expected = expected[:tt.delivery]
```
Comment (powersj) [inline plugins/processors/starlark/starlark_test.go:3543]: Why are you taking the subslice here?

Comment (srebhan) [reply plugins/processors/starlark/starlark_test.go:3543]: The `expected` array is filled with the tracking information of the input data, however, we might want to "hold" some of the metrics in the starlark script, so not all input data reaches the output, like in the second test-cases, where one metric is held in the starlark processor and thus cannot be acknowledged by the output **and** the processor as required due to the `deepcopy` (i.e. `refcount=2`).

Comment (powersj) [reply plugins/processors/starlark/starlark_test.go:3543]: ok then `delivery` is the number of deliveries we expect to see at the end?

Comment (srebhan) [reply plugins/processors/starlark/starlark_test.go:3543]: Exactly. It's a poor naming again from my side I guess... :-(

**Review by powersj — APPROVED** [review]

**Pull request merged**